enhancements/20230830-managed-dhcp.md

```diff
@@ -0,0 +1,318 @@
+
+### Test plan
+
+#### Private VM Networks
+
+The plan is to create two VM networks that are isolated and separate from the outside. The VM attached to these two networks will be assigned IP addresses for each NIC from the DHCP services and gain network connectivity (within these two networks).
+
+Prepare a Harvester cluster consists of multiple nodes (2+). We'll need to create two cluster networks so each Harvester node should have more than two NICs.
+
+1. Install the `vm-dhcp-controller` chart
+1. Create a cluster network called `provider-all` (the cluster network consists of **all the nodes**)
+1. Create a network config called `provider-all-nc` for the cluster network `provider-all` with all nodes and one of their NICs as the uplink
+1. Create another cluster network called `provider-cp` (the cluster network consists of **only the control plane nodes**)
+1. Create a network config called `provider-cp-nc` for the cluster network `provider-cp` with only the control plane nodes and one of their NICs as the uplink
+1. Create a VM network called `priv-net-all` with cluster network `provider-all` selected. The VLAN ID should be set with a number which is not in use in the premises, say, `817`.
+1. Create another VM network called `priv-net-cp` with cluster network `provider-cp` selected. The VLAN ID should be set with a number which is not in use in the premises and the above VM network, say, `689`.
+1. Create a VM called `test-vm` with two NICs attaching to the `priv-net-all` and the `priv-net-cp` networks respectively. Also, the NIC's MAC address for `priv-net-all` is `02:d4:6f:a8:bc:ed` and the other for `priv-net-cp` is `24:ab:56:78:9c:01`.
+
+From now on, the steps will interact with the managed DHCP feature.
+
+1. Create an IPPool object `priv-net-all` with the following YAML content:
+
+    ```yaml
+    apiVersion: network.harvesterhci.io/v1alpha1
+    kind: IPPool
+    metadata:
+      name: priv-net-all
+      namespace: default
+    spec:
+      ipv4Config:
+        serverIP: 172.19.150.2
+        cidr: 172.19.150.0/28
+        pool:
+        start: 172.19.150.5
+        end: 172.19.150.14
+        leaseTime: 600
+      networkName: default/priv-net-all
+    ```
+
+2. Create another IPPool object `priv-net-cp` with the following YAML content:
+
+    ```yaml
+    apiVersion: network.harvesterhci.io/v1alpha1
+    kind: IPPool
+    metadata:
+      name: priv-net-cp
+      namespace: default
+    spec:
+      ipv4Config:
+        serverIP: 172.19.100.2
+        cidr: 172.19.100.0/28
+        pool:
+        start: 172.19.100.10
+        end: 172.19.100.14
+        leaseTime: 300
+      networkName: default/priv-net-cp
+    ```
+
+3. Create a VirtualMachineNetworkConfig object for the VM `test-vm-1`:
+
+    ```yaml
+    apiVersion: network.harvesterhci.io/v1alpha1
```

Comment:
The `VirtualMachineNetworkConfig` looks to be redundant with kubevirt vm config `spec.domain.devices.interfaces`, does the user need to config vm MAC address in two places?

```
type Interface struct {
	// Logical name of the interface as well as a reference to the associated networks.
	// Must match the Name of a Network.
	Name string `json:"name"`
	// Interface model.
	// One of: e1000, e1000e, ne2k_pci, pcnet, rtl8139, virtio.
	// Defaults to virtio.
	// TODO:(ihar) switch to enums once opengen-api supports them. See: https://github.com/kubernetes/kube-openapi/issues/51
	Model string `json:"model,omitempty"`
	// BindingMethod specifies the method which will be used to connect the interface to the guest.
	// Defaults to Bridge.
	InterfaceBindingMethod `json:",inline"`
	// List of ports to be forwarded to the virtual machine.
	Ports []Port `json:"ports,omitempty"`
	// Interface MAC address. For example: de:ad:00:00:be:af or DE-AD-00-00-BE-AF.
	MacAddress string `json:"macAddress,omitempty"`
	// BootOrder is an integer value > 0, used to determine ordering of boot devices.
	// Lower values take precedence.
	// Each interface or disk that has a boot order must have a unique value.
	// Interfaces without a boot order are not tried.
	// +optional
```

A straight-forward idea is, config the `macaddress` in vm spec; config `mac-ip` binding in the pool, then the VM can get the desired ip from the pool.

Reply:
Configuring the MAC address for both the VM and the vmnetcfg object is unnecessary, as the controller will monitor the VMI object to retrieve the VM's MAC address and use it to create the corresponding vmnetcfg object. This is the dynamic-bind scenario.

For the static-bind scenario, it will need a deeper integration with Harvester. This is tracked in #5052.